FuelLabs/sway
# Prevent failure while creating BuildPlan from lock file in the case of a dependency removal

## About this PR
~Still in WIP, opening up for visibility. I will be adding testing steps~

- [x]  Merge #1836 first and add required patching checks to this one as well

closes #1778.

### Motivation
Before this PR, `BuildPlan` generation from the lock file was failing if after the lock file is generated, a dependency is removed from the manifest file (specifically if the removed dependency is a path dependency).

This was resulting in the `BuildPlan` being generated from scratch. To generate the `BuildPlan` from scratch we generally need connectivity since `std` is implicitly added as a git dependency (if we do not explicitly prevent it with `implicit-std = false`). Ideally, we should be able to remove any dependency in offline mode. 

With this PR, we should be able to create the `BuildPlan` from the lock file even if a path dependency is removed. This *should* mean `apply_pkg_diff` will be able to remove any dependency in `offline` mode.


### To Test

1. run `forc new parent`
2. run `forc new --library child`
3. add child as a dependency to parent
4. run `forc build` in parent
5. remove the child from parent's dependencies
6. run `forc build` in parent

This should output something like

```
Creating a new `Forc.lock` file. (Cause: lock file did not match manifest `diff`)
```

Rather than 

```
Creating a new `Forc.lock` file. (Cause: dependency required for path reconstruction has been removed from the manifest)
```
